src/NzbDrone.Core/DecisionEngine/Specifications/UpgradableSpecification.cs

```diff
@@ -157,5 +244,45 @@ public bool IsRevisionUpgrade(QualityModel currentQuality, QualityModel newQuali
 
             return false;
         }
+
+        public bool IsUpgradeAllowed(QualityProfile qualityProfile, LanguageProfile languageProfile, List<QualityModel> currentQualities, List<Language> currentLanguages, QualityModel newQuality, Language newLanguage)
+        {
+            var isQualityUpgrade = IsQualityUpgradable(qualityProfile, currentQualities, newQuality);
+            var isLanguageUpgrade = IsLanguageUpgradable(languageProfile, currentLanguages, newLanguage);
+
+            return CheckUpgradeAllowed(qualityProfile,languageProfile,isQualityUpgrade,isLanguageUpgrade);
+        }
+
+        public bool IsUpgradeAllowed(QualityProfile qualityProfile, LanguageProfile languageProfile, QualityModel currentQuality, Language currentLanguage, QualityModel newQuality, Language newLanguage)
+        {
+            var isQualityUpgrade = new QualityModelComparer(qualityProfile).Compare(newQuality, currentQuality) > 0;
+            var isLanguageUpgrade = new LanguageComparer(languageProfile).Compare(newLanguage, currentLanguage) > 0;
+
+            return CheckUpgradeAllowed(qualityProfile, languageProfile, isQualityUpgrade, isLanguageUpgrade);
+        }
+
+        private bool CheckUpgradeAllowed (QualityProfile qualityProfile, LanguageProfile languageProfile, bool isQualityUpgrade, bool isLanguageUpgrade)
+        {
+            if (isQualityUpgrade && qualityProfile.UpgradeAllowed ||
+                isLanguageUpgrade && languageProfile.UpgradeAllowed)
+            {
+                _logger.Debug("At least one profile allows upgrading");
+                return true;
+            }
+
+            if (isQualityUpgrade && !qualityProfile.UpgradeAllowed)
+            {
+                _logger.Debug("Quality profile allows upgrades, skipping");
```

Comment:
I think this should read `Quality profile does not allow upgrades, skipping` (see case below)

Reply:
Yep you are right, I had pointed that out to Mark on Sonarr but forgot to change it myself... what can you do :)